lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -1437,22 +1456,37 @@ void cudaq::cc::ReturnOp::getCanonicalizationPatterns(
 }
 
 //===----------------------------------------------------------------------===//
-// ConditionOp
+// SizeOp
 //===----------------------------------------------------------------------===//
 
-LogicalResult cudaq::cc::ConditionOp::verify() {
-  Operation *self = getOperation();
-  Region *region = self->getBlock()->getParent();
-  auto parentOp = self->getParentOfType<LoopOp>();
-  assert(parentOp); // checked by tablegen constraints
-  if (&parentOp.getWhileRegion() != region)
-    return emitOpError("only valid in the while region of a loop");
-  return success();
-}
+namespace {
+struct ReplaceConstantSizes : public OpRewritePattern<cudaq::cc::SizeOfOp> {
+  using Base = OpRewritePattern<cudaq::cc::SizeOfOp>;
+  using Base::Base;
 
-MutableOperandRange cudaq::cc::ConditionOp::getMutableSuccessorOperands(
-    std::optional<unsigned> index) {
-  return getResultsMutable();
+  LogicalResult matchAndRewrite(cudaq::cc::SizeOfOp sizeOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO: Add handling of more types.
+    auto inpTy = sizeOp.getInputType();
+    std::optional<unsigned> bitsize;
+    if (isa<IntegerType, FloatType>(inpTy)) {
+      bitsize = inpTy.getIntOrFloatBitWidth();
+    } else if (auto strTy = dyn_cast<cudaq::cc::StructType>(inpTy)) {
+      if (auto bits = strTy.getBitSize())
+        bitsize = bits;
+    }
+    if (!bitsize)
+      return failure();
```

Comment:
Will this show a nice user error message if the type isn't supported yet? I.e., what would happen if they try to take the sizeof a qubit?

Edit: Actually, will this new operator get invoked if the user adds `sizeof` to their code? I'm guessing that since it's not in the bridge, then it won't happen.

Reply:
This is a quirk of the MLIR interface, but "failure" in the canonicalization pattern means "we don't have anything here to update" rather than "the train went over the cliff".

`sizeof qubit` is nonsense and will be caught in codegen by generating a `poison` value. So don't do it.